File: internal/app/reporter.go
package app

type StatusKind string

const (
	StatusOK      StatusKind = "ok"
	StatusStale   StatusKind = "stale"
	StatusMissing StatusKind = "missing"
)

type ProgressReporter interface {
	Increment(label string)
	Done()
}

type Reporter interface {
	Info(message string)
	Status(kind StatusKind, source, output, lang string)
	StatusSummary(ok, stale, missing int)
	CleanRemoved(path string)
	CleanMissing(path string)
	CleanSummary(removed, missing, lockRemoved int)
	Progress(label string, total int) ProgressReporter
}

type noopReporter struct{}

func (n noopReporter) Info(string)                               {}
func (n noopReporter) Status(StatusKind, string, string, string) {}
func (n noopReporter) StatusSummary(int, int, int)               {}
func (n noopReporter) CleanRemoved(string)                       {}
func (n noopReporter) CleanMissing(string)                       {}
func (n noopReporter) CleanSummary(int, int, int)                {}
func (n noopReporter) Progress(string, int) ProgressReporter     { return noopProgress{} }

type noopProgress struct{}

func (n noopProgress) Increment(string) {}
func (n noopProgress) Done()            {}

func ensureReporter(reporter Reporter) Reporter {
	if reporter == nil {
		return noopReporter{}
	}
	return reporter
}

File: src/reporter.ts
export type Verb =
  | "Translating"
  | "Validating"
  | "Checking"
  | "Ok"
  | "Stale"
  | "Missing"
  | "Removed"
  | "Skipped"
  | "Translated"
  | "Cleaned"
  | "Created"
  | "Updated"
  | "Summary"
  | "Info"
  | "Dry run";

export interface ProgressReporter {
  increment(label: string): void;
  done(): void;
}

export interface Reporter {
  log(verb: Verb, message: string): void;
  step(verb: Verb, current: number, total: number, message: string): void;
  blank(): void;
  progress(verb: Verb, total: number): ProgressReporter;
}

class NoopProgress implements ProgressReporter {
  increment(_label: string): void {}
  done(): void {}
}

class NoopReporter implements Reporter {
  log(_verb: Verb, _message: string): void {}
  step(_verb: Verb, _current: number, _total: number, _message: string): void {}
  blank(): void {}
  progress(_verb: Verb, _total: number): ProgressReporter {
    return new NoopProgress();
  }
}

export function ensureReporter(reporter?: Reporter): Reporter {
  return reporter ?? new NoopReporter();
}
